### PR TITLE
feat: add decimal validation for numeric precision and scale supported by Spanner

### LINKED ISSUE #339: Add validation in decimal to numeric conversion
Add validation in decimal to numeric conversion to ensure precision and recall is supported by spanner.

read more here: https://cloud.google.com/spanner/docs/data-types#numeric_type.

Similar validations are present in other clients eg: https://github.com/googleapis/java-spanner/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java#L141-L155

### PR DESCRIPTION
**Pull request by vi3k6i5**
feat: add decimal validation for numeric precision and scale supported by spanner
Fixes #339 🦕

### CODE REVIEW

File: google/cloud/spanner_dbapi/parse_utils.py
Comment (olavloite) [inline google/cloud/spanner_dbapi/parse_utils.py:527]: I don't know whether this is normal in the Python client, but it seems a little strange that this function is called for __all__ parameter types, and not only numeric values. I would have expected either:
1. That this function is only called for numeric values (i.e. the `if isinstance(value, decimal.Decimal)` would be outside of the function).
2. Or: That there is a generic 'assert_valid_value(value)` function that first checks the type of the input parameter, and then calls the appropriate validation function for that type.

**Review by olavloite — APPROVED** [review]
This seems good to my untrained Python eyes, with a small question on whether it is logical to call the validate function for all values, or whether it would be better to move the if-check for the type out of the validation function.
(You should in any case await approval from other reviewers before merging)

File: google/cloud/spanner_dbapi/parse_utils.py
Comment (larkee) [inline google/cloud/spanner_dbapi/parse_utils.py:527]: I think this logic would make more sense to be added to the base client [here](https://github.com/googleapis/python-spanner/blob/master/google/cloud/spanner_v1/_helpers.py#L131-L132).

Comment (vi3k6i5) [reply google/cloud/spanner_dbapi/parse_utils.py:527]: makes sense added an instance check before the method is called.

Comment (vi3k6i5) [reply google/cloud/spanner_dbapi/parse_utils.py]: moved it to base client.

**Comment by vi3k6i5** [comment]
> This seems good to my untrained Python eyes, with a small question on whether it is logical to call the validate function for all values, or whether it would be better to move the if-check for the type out of the validation function.
> (You should in any case await approval from other reviewers before merging)

moved the method after a check. and also moved it to base client as @larkee suggested.

File: google/cloud/spanner_v1/_helpers.py
Comment (larkee) [inline google/cloud/spanner_v1/_helpers.py:107]:
```suggestion
    :type value: decimal.Decimal
```

File: google/cloud/spanner_v1/_helpers.py
Comment (larkee) [inline google/cloud/spanner_v1/_helpers.py:106]:
```suggestion
    Asserts that input numeric field is within Spanner supported range.
```
Also this should be moved to the top of the docstring. That way you have an explanation of what function does, followed by why it's needed.

File: google/cloud/spanner_v1/_helpers.py
Comment (larkee) [inline google/cloud/spanner_v1/_helpers.py:110]:
```suggestion
    :raises NotSupportedError: if value is not within supported precision or scale of Spanner.
```

File: tests/unit/test__helpers.py
Comment (larkee) [inline tests/unit/test__helpers.py:250]: I don't think this line is needed. The value should be coming from the cases.

File: google/cloud/spanner_v1/_helpers.py
Comment (vi3k6i5) [reply google/cloud/spanner_v1/_helpers.py:107]: fixed.

File: google/cloud/spanner_v1/_helpers.py
Comment (vi3k6i5) [reply google/cloud/spanner_v1/_helpers.py:106]: Done.

File: tests/unit/test__helpers.py
Comment (vi3k6i5) [reply tests/unit/test__helpers.py:250]: changed to read from cases.

File: google/cloud/spanner_v1/_helpers.py
Comment (vi3k6i5) [reply google/cloud/spanner_v1/_helpers.py:110]: fixed comment.

**Review by larkee — APPROVED** [review]
LGTM :+1:  You could remove `test_w_numeric` if you wanted since `test_w_numeric_precision_and_scale_valid` should have more coverage.

**Pull request merged**
